edb/tools/gen_rust_ast.py

```diff
@@ -0,0 +1,274 @@
+from edb.common import enum as s_enum
+from edb.tools.edb import edbcommands
+
+
+@dataclasses.dataclass()
+class ASTClass:
+    name: str
+    typ: typing.Type
+    children: typing.List[typing.Type] = dataclasses.field(default_factory=list)
+
+
+@dataclasses.dataclass()
+class ASTUnion:
+    name: str
+    variants: typing.Sequence[typing.Type | str]
+    for_composition: bool
+
+
+# a queue for union types that are to be generated
+union_types: typing.List[ASTUnion] = []
+
+# all discovered AST classes
+ast_classes: typing.Dict[str, ASTClass] = {}
+
+
+@edbcommands.command("gen-rust-ast")
+def main() -> None:
+    f = open('edb/edgeql-parser/src/ast.rs', 'w')
+
+    f.write(
+        textwrap.dedent(
+            '''\
+            // DO NOT EDIT. This file was generated with:
+            //
+            // $ edb gen-rust-ast
+
+            //! Abstract Syntax Tree for EdgeQL
+            #![allow(non_camel_case_types)]
+
+            use std::collections::HashMap;
+
+            #[cfg(feature = "python")]
+            use edgeql_parser_derive::IntoPython;
+            '''
+        )
+    )
+
+    # discover all nodes
+    for name, typ in qlast.__dict__.items():
+        if not isinstance(typ, type) or not hasattr(typ, '_direct_fields'):
+            continue
+
+        if hasattr(typ, '__rust_ignore__') and typ.__rust_ignore__:
+            continue
+
+        # re-run field collection to correctly handle forward-references
+        typ = typ._collect_direct_fields()  # type: ignore
+
+        ast_classes[typ.__name__] = ASTClass(name=name, typ=typ)
+
+    # build inheritance graph
+    for ast_class in ast_classes.values():
+        for base in ast_class.typ.__bases__:
+            if base.__name__ not in ast_classes:
+                continue
+            ast_classes[base.__name__].children.append(ast_class.typ)
+
+    # generate structs
+    for ast_class in ast_classes.values():
+        f.write(codegen_struct(ast_class))
+
+        while len(union_types) > 0:
+            f.write(codegen_union(union_types.pop(0)))
+
+    # generate enums
+    for name, typ in chain(qlast.__dict__.items(), qltypes.__dict__.items()):
+
+        if not isinstance(typ, type) or not issubclass(typ, s_enum.StrEnum):
+            continue
+
+        f.write(codegen_enum(name, typ))
+
+
+def codegen_struct(cls: ASTClass) -> str:
+    field_names = set()
+    fields = ''
+    doc_comment = ''
+
+    for f in typing.cast(typing.List[ast._Field], cls.typ._direct_fields):
+
+        if f.hidden:
+            continue
+
+        union_name = f'{cls.name}{title_case(f.name)}'
+
+        typ = translate_type(f.type, union_name, False)
+        if hasattr(cls.typ, '__rust_box__') and f.name in cls.typ.__rust_box__:
+            typ = f'Box<{typ}>'
+
+        f_name = quote_rust_ident(f.name)
+        field_names.add(f_name)
+
+        fields += f'    pub {f_name}: {typ},\n'
+
+    if len(cls.children) > 0:
+
+        for i in range(0, 10):
```

Comment:
using `range(0, 10)` is hacky; try `itertools.count` if you don't want to write a while loop